include/proteus/Frontend/Func.hpp

```diff
@@ -836,23 +865,58 @@ Var<T, std::enable_if_t<std::is_array_v<T>>>::operator[](
 
 template <typename T>
 Var<std::remove_pointer_t<T>>
-Var<T, std::enable_if_t<std::is_pointer_v<T>>>::operator[](size_t Index) {
+Var<T, std::enable_if_t<std::is_pointer_v<T>>>::indexImpl(Value *IdxValue) {
   auto &IRB = Fn.getIRBuilder();
 
-  auto *PointerElemTy = getValueType();
-  auto *Ptr = loadPointer();
-  auto *GEP = IRB.CreateConstInBoundsGEP1_64(PointerElemTy, Ptr, Index);
-  unsigned AddrSpace = cast<PointerType>(Ptr->getType())->getAddressSpace();
-  Type *ElemPtrTy = PointerType::get(PointerElemTy, AddrSpace);
+  if (!IdxValue->getType()->isIntegerTy())
+    PROTEUS_FATAL_ERROR("Pointer index must be an integer type");
 
-  // Create a pointer storage to hold the LValue for
-  // the Array[Index].
-  auto *PtrSlot = Fn.emitAlloca(ElemPtrTy, "elem.ptr");
-  IRB.CreateStore(GEP, PtrSlot);
-  std::unique_ptr<VarStorage> ResultStorage =
-      std::make_unique<PointerStorage>(PtrSlot, IRB, PointerElemTy);
+  if (IdxValue->getType()->getIntegerBitWidth() != 64)
+    IdxValue = IRB.CreateSExtOrTrunc(IdxValue, IRB.getInt64Ty());
+
+  auto *PointerElemTy = getValueType();
+  auto *AddressBase = loadPointer();
+  unsigned AddrSpaceAddr =
+      cast<PointerType>(AddressBase->getType())->getAddressSpace();
+
+  using RetElemT = std::remove_pointer_t<T>;
+  auto &Ctx = IRB.getContext();
+  if constexpr (std::is_pointer_v<RetElemT>) {
```

Comment:
When does this happen? Wouldn't the specialization `std::is_pointer_v` have taken over?